src/main/java/org/craftercms/studio/impl/v2/event/EventBroadcaster.java

```diff
@@ -52,17 +53,10 @@ public void publishGlobalEvent(final GlobalBroadcastEvent event) {
         publishEvent(event, DESTINATION_ROOT);
     }
 
+    @LogExecutionTime
     private void publishEvent(final BroadcastEvent event, final String destination) {
```

Comment:
Does it work with private methods?

Reply:
@jmendeza It does not work with private methods. Do we have any concerns if I update these methods to public?

Reply:
@phuongnq I believe `protected` should do, let's try that first

Reply:
@jmendeza It seems we are using Spring AOP Proxies and the annotation does not trigger on the internal method call regardless it is public/protected/private method: https://docs.spring.io/spring-framework/reference/core/aop/proxying.html#aop-understanding-aop-proxies
For this particular method, I have moved the `@LogExecutionTime` to the upper public method.
For other places, I have reverted the change and I'm currently looking for other solutions. Please let me know if you have any recommendations! 

We can for example have a lambda to wrap the function like this:
```
TimeUtils.logExecutionTime(() -> method(param1, param2));
```
But I found it is not very clean.

Another solution such as using load-time weaving with AspectJ: https://docs.spring.io/spring-framework/reference/core/aop/using-aspectj.html#aop-aj-ltw but this will make some structure change required.

Reply:
I see. Let's try load-time weaving. We should measure differences in startup time and memory footprint though.
Can we do compile-time weaving?

Reply:
Regarding compile-time weaving, I'm not optimistic with our project yet. We will need to add the following plugin to `pom.xml`:

```
<plugin>
				<groupId>org.codehaus.mojo</groupId>
				<artifactId>aspectj-maven-plugin</artifactId>
				<version>1.15.0</version>
				<executions>
					<execution>
						<goals>
							<goal>compile</goal>       <!-- use this goal to weave all your main classes -->
							<goal>test-compile</goal>  <!-- use this goal to weave all your test classes -->
						</goals>
					</execution>
				</executions>
				<configuration>
					<source>16</source>
					<target>16</target>
					<complianceLevel>17.0</complianceLevel>
					<showWeaveInfo>true</showWeaveInfo>
					<encoding>UTF-8</encoding>
				</configuration>
			</plugin>
```
But while run `mvn clean install`, I got some error:
```
[INFO] Showing AJC message detail for messages of types: [error, warning, fail]
[WARNING] bad version number found in /home/pnguyen/.m2/repository/org/aspectj/aspectjrt/1.9.21/aspectjrt-1.9.21.jar expected 1.9.7 found 1.9.21
	<unknown source file>:<no line information>

[ERROR] around on staticinitialization of interface 'org.craftercms.studio.api.v2.annotation.RequireSiteReady' not supported (compiler limitation)
	/home/pnguyen/craftercms/next/craftercms/src/studio/src/main/java/org/craftercms/studio/api/v2/annotation/RequireSiteStateAnnotationHandler.java:58

[ERROR] around on staticinitialization of interface 'org.craftercms.studio.api.v2.annotation.RequireSiteReady' not supported (compiler limitation)
	/home/pnguyen/craftercms/next/craftercms/src/studio/src/main/java/org/craftercms/studio/api/v2/annotation/RequireSiteReady.java:35
public @interface RequireSiteReady {
                  ^^^^^^^^^^^^^^^

[ERROR] around on staticinitialization of interface 'org.craftercms.studio.api.v2.annotation.RequireContentExists' not supported (compiler limitation)
	/home/pnguyen/craftercms/next/craftercms/src/studio/src/main/java/org/craftercms/studio/api/v2/annotation/RequireSiteExistsAnnotationHandler.java:60

[ERROR] around on staticinitialization of interface 'org.craftercms.studio.api.v2.annotation.RequireContentExists' not supported (compiler limitation)
	/home/pnguyen/craftercms/next/craftercms/src/studio/src/main/java/org/craftercms/studio/api/v2/annotation/RequireContentExists.java:27
public @interface RequireContentExists {
                  ^^^^^^^^^^^^^^^^^^^

[INFO] Join point 'method-execution(org.craftercms.studio.model.rest.ResponseBody org.craftercms.studio.controller.rest.v2.ConfigurationController.getConfiguration(java.lang.String, java.lang.String, java.lang.String, java.lang.String))' in Type 'org.craftercms.studio.controller.rest.v2.ConfigurationController' (ConfigurationController.java:88) advised by around advice from 'org.craftercms.studio.api.v2.annotation.LogExecutionTimeAnnotationHandler' (LogExecutionTimeAnnotationHandler.java:50)
[INFO] Join point 'method-execution(org.craftercms.studio.model.rest.ResponseBody org.craftercms.studio.controller.rest.v2.ModelController.getModelDefinitions(java.lang.String))' in Type 'org.craftercms.studio.controller.rest.v2.ModelController' (ModelController.java:52) advised by around advice from 'org.craftercms.studio.api.v2.annotation.LogExecutionTimeAnnotationHandler' (LogExecutionTimeAnnotationHandler.java:50)
[ERROR] The package org.xml.sax is accessible from more than one module: <unnamed>, java.xml
	/home/pnguyen/craftercms/next/craftercms/src/studio/src/main/java/org/craftercms/studio/impl/v1/util/ContentUtils.java:25
import org.xml.sax.SAXException;
       ^^^^^^^^^^
```
This could due to the plugin is not fully supporting Java 17 or/and needing some more configuration where we do the compile/where to not.
Note that I also tried the other plugin as some suggestion in the above plugin github but the result is the same: https://github.com/dev-aspectj/aspectj-maven-plugin

Reply:
cc @sumerjabri

Reply:
@phuongnq I'm curious why is <target> and <source> set to `16` in the plugin configuration

But anyway, that path seems more complex than expected and it would be adding more moving pieces, i.e. future maintenance. Not a great trade-off in my opinion.
I believe having some helper method like the suggested is an acceptable compromise for the cases where the annotation won't work:
```
TimeUtils.logExecutionTime(() -> method(param1, param2));
```
@sumerjabri

Reply:
Updated to have additional `TimeUtils.logExecutionTime()` method.
This lambda also has some limitations in that we should not run it against functions with exceptions in the signature because we then have to try/catch the RunTimeException inside the lambda and it could mean loss of the original exception instance.

For new implementing methods now on, I think we should design them so that it is `public` and external call to reuse the annotation. That would make the code clean.

FYI
> I'm curious why are target and source set to 16 in the plugin configuration

This is due to this commit: https://github.com/mojohaus/aspectj-maven-plugin/pull/155/files#diff-a03bee29a28d152f1791ed2a8b35c38152dee87a5d429604007ccd748c417544. For some reason, they use target and source 16 to support java 17.

Reply:
@jmendeza, we need to document somewhere for us and future developers how to weave this into the methods to trace execution time cleanly, as Phil indicated.